File: src/algorithms/binarySearch.ts
function binarySearch(
  input: string[],
  target: string,
  low: number = 0,
  hight: number = input.length
): number | null {
  if (input.length === 0) {
    return null;
  } else if (input.length === 1) {
    return input[0] === target ? 0 : null;
  }

  const averageIndex = Math.round((low + hight) / 2);
  const averageValue = input[averageIndex];

  if (averageValue === target) {
    return averageIndex;
  }

  const newBorders = target > averageValue ? [averageIndex, hight] : [0, averageIndex];

  return binarySearch(input, target, ...newBorders)
}

export default binarySearch

File: grokking-algorithms-python/binary_search.py
# Chapter 1 | Introduction to algorithms


def binary_search(input, value, low, high):
    mid = round((low + high) / 2)
    if input[mid] == value:
        return mid
    if input[mid] < value:
        return binary_search(input, value, mid + 1, high)
    return binary_search(input, value, low, mid - 1)
